app/assets/javascripts/beak/widgets/ractives/code-editor.coffee

```diff
@@ -49,9 +50,48 @@ window.RactiveModelCodeComponent = Ractive.extend({
     $('#procedurenames-dropdown').trigger('chosen:updated')
     return
 
+  setupAutoComplete: (hintList) ->
+    CodeMirror.registerHelper('hintWords', 'netlogo', hintList)
+    editor = @findComponent('codeEditor').getEditor()
+    editor.on('keyup', (cm, event) =>
+      if not cm.state.completionActive and event.keyCode > 64 and event.keyCode < 91 and @get('autoCompleteStatus')
+        cm.showHint({completeSingle: false})
+    )
+    return
+
+  netLogoHintHelper: (cm, options) ->
+    cur = cm.getCursor()
+    token = cm.getTokenAt(cur)
+    to = CodeMirror.Pos(cur.line, token.end)
+    if token.string and /\S/.test(token.string[token.string.length - 1])
+      term = token.string
+      from = CodeMirror.Pos(cur.line, token.start)
+    else
+      term = ''
+      from = to
+    found = options.words.filter( (word) ->
+      if word.slice(0, term.length) is term
+        word
+    )
```

Comment:
This inner function does not strictly follow `filter`'s type signature.  `filter` is of type `((A) => Boolean) => Array[A]`.  So the function you give it should be of type `(A) => Boolean`.  But the function you're providing is of type `(A) => (A|Unit)` (because, when `if` succeeds, you return the word (`A`), and when it doesn't, you return nothing `null`/`Unit`).  This _works_, because JavaScript does weird things with treating `null` as `false` and treating "truthy" things (like `A`s) as `true`.  But doing this makes the code longer and more difficult to reason about.  This can be expressed more simply with `found = options.words.filter((word) -> word.slice(0, term.length) is term)`, which will be brief, clear (to me, at least), and rigorously typed.